src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,326 @@
+    : AsyncContextFrame(IsolateBase::from(js.v8Isolate)) {
+
+  // Propagate the storage context of the parent frame to this newly created frame.
+  const auto propagate = [&](AsyncContextFrame& parent) {
+    parent.storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+    for (auto& entry : parent.storage) {
+      storage.insert(entry.clone(js));
+    }
+
+    KJ_IF_MAYBE(entry, maybeStorageEntry) {
+      storage.upsert(kj::mv(*entry), [](StorageEntry& existing, StorageEntry&& row) mutable {
+        existing.value = kj::mv(row.value);
+      });
+    }
+  };
+
+  KJ_IF_MAYBE(parent, maybeParent) {
+    propagate(*parent);
+  } else {
+    propagate(current(js));
+  }
+}
+
+kj::Maybe<AsyncContextFrame&> AsyncContextFrame::tryGetContext(
+    Lock& js,
+    v8::Local<v8::Promise> promise) {
+  auto handle = js.getPrivateSymbolFor(Lock::PrivateSymbols::ASYNC_RESOURCE);
+  // We do not use the normal unwrapOpaque here since that would consume the wrapped
+  // value, and we need to be able to unwrap multiple times.
+  return tryGetContextFrame(js.v8Isolate,
+      check(promise->GetPrivate(js.v8Isolate->GetCurrentContext(), handle)));
+}
+
+kj::Maybe<AsyncContextFrame&> AsyncContextFrame::tryGetContext(
+    Lock& js,
+    V8Ref<v8::Promise>& promise) {
+  return tryGetContext(js, promise.getHandle(js));
+}
+
+AsyncContextFrame& AsyncContextFrame::current(Lock& js) {
+  auto& isolateBase = IsolateBase::from(js.v8Isolate);
+  if (isolateBase.asyncFrameStack.size() == 0) {
+    return *isolateBase.getRootAsyncContext();
+  }
+  return *isolateBase.asyncFrameStack.back();
+}
+
+Ref<AsyncContextFrame> AsyncContextFrame::create(
+    Lock& js,
+    kj::Maybe<AsyncContextFrame&> maybeParent,
+    kj::Maybe<StorageEntry> maybeStorageEntry) {
+  return alloc<AsyncContextFrame>(js, maybeParent, kj::mv(maybeStorageEntry));
+}
+
+v8::Local<v8::Function> AsyncContextFrame::wrap(
+    Lock& js,
+    v8::Local<v8::Function> fn,
+    kj::Maybe<AsyncContextFrame&> maybeFrame,
+    kj::Maybe<v8::Local<v8::Value>> thisArg) {
+  auto isolate = js.v8Isolate;
+  auto context = isolate->GetCurrentContext();
+  return js.wrapReturningFunction(context, JSG_VISITABLE_LAMBDA(
+      (
+        frame = kj::addRef(AsyncContextFrame::current(js)),
+        thisArg = js.v8Ref(thisArg.orDefault(context->Global())),
+        fn = js.v8Ref(fn)
+      ),
+      (thisArg, fn),
+      (Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+    auto function = fn.getHandle(js);
+    auto context = js.v8Isolate->GetCurrentContext();
+
+    kj::Vector<v8::Local<v8::Value>> argv(args.Length());
+    for (int n = 0; n < args.Length(); n++) {
+      argv.add(args[n]);
+    }
+
+    AsyncContextFrame::Scope scope(js, *frame);
+    v8::Local<v8::Value> result;
+    if (function->Call(context, thisArg.getHandle(js), args.Length(), argv.begin())
+            .ToLocal(&result)) {
+      return result;
+    }
+    return js.v8Undefined();
```

Comment:
If `function->Call()` return an empty handle, that means an exception was thrown, which is different from returning undefined.

I think you want this:

```suggestion
    return jsg::check(function->Call(context, thisArg.getHandle(js), args.Length(), argv.begin()));
```

And we should add a test for the case where a wrapped function throws an exception.

Reply:
As a side note... adding a `js.call(...)` utility to `jsg::Lock` might be nice to eliminate some of the boilerplate necessary for calling js functions directly.